language_bindings/rust/rust_build/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+        }
+    }
+    /// Override the output directory.
+    ///
+    /// by default this is the ENV "out_dir" but can be anywere
+    /// useful to place .rs and headers in a folder in the src of your project but doing that
+    /// means cargo does not clean these artifacts
+    pub fn out_dir<T: Into<PathBuf>>(mut self, out: T) -> Self {
+        self.rs_output = out.into();
+        self
+    }
+    /*
+    pub fn push_gen(mut self, gen: Generator<'static>)->Self{
+        self.generators.push(gen);
+        self
+    }
+
+     */
+    //Todo template all strings
+    /// Create a generator based on its name
+    ///
+    /// We have strict naming conventions:   
+    ///    - The generator source must be gen_name.cpp   
+    ///    - The internal cpp class must also be gen_name
+    ///
+    ///```ignore
+    /// let gen = genBuilder.new_gen(iir_blur);
+    /// ```
+    pub fn new_gen(self, gen_name: String) -> Generator<'static> {
+        let mut temp_gen_exe = PathBuf::from(self.rs_output.to_str().unwrap());
+        if gen_name.is_empty() {
+            temp_gen_exe.push("runtime");
+        } else {
+            println!(
+                "cargo:rerun-if-changed={}",
+                self.gen_path
+                    .join(gen_name.as_str())
+                    .with_extension("cpp")
+                    .to_str()
+                    .unwrap()
+            );
+            temp_gen_exe.push(gen_name.as_str());
+        }
+        temp_gen_exe.with_extension("generator");
+
+        println!("temp: {:?}", temp_gen_exe.as_os_str());
+
+        Generator {
+            gen_name: gen_name.to_string(),
+            gen_exe: temp_gen_exe,
+            halide_path: self.halide_path.clone(),
+            gen_path: self.gen_path.join(gen_name).with_extension("cpp"),
+            rs_output: self.rs_output.clone(),
+            gcc_flags: vec!["-lHalide", "-ldl", "-lpthread", "-lz"], //Todo add adders
+            debug: self.debug,                                       //Todo add functionality
+            target: self.target,                                     //todo add setter
+        }
+    }
+    /// Sets the console debug for Halide generators by default this is false
+    pub fn debug(mut self, b: bool) -> Self {
+        self.debug = b;
+        self
+    }
+    ///
+    ///
+    ///
+    pub fn make_runtime(self) {
+        let gen = self.new_gen("".to_string());
+
+        let compilation_result = gen.compile();
+        println!("Gen Creation Status: {}", compilation_result.status);
+        assert!(compilation_result.status.success());
+
+        let execution_result = gen.make_runtime();
+        println!("runtime results: {:?}", execution_result);
+        assert!(execution_result.is_ok());
+
+        gen.rename_runtime();
+    }
+}
+
+impl Generator<'static> {
+    ///Make the generator executable using Halide GenGen and g++
+    pub fn compile(&self) -> Output {
+        let mut cmd_compile = Command::new("g++");
```

Comment:
For portability, this should use the `cc` crate rather than hard-coding a dependency on g++ specifically.

Reply:
We attempted to use CC crate but found that it was automatically adding lib and .a to the compilation, could not find a workaround here. To be clear this cmd_complie makes an executable that is run later. If you are aware of a workaround here to use CC we would prefer that as well.

Reply:
I'm not familiar with the Halide workflow. Can you clarify for me why we're trying to compile an executable binary here? I thought the goal was to compile a static library that downstream Rust code can link to. Could the binary be instead be obtained from a precompiled Halide installation?